File: javapatterns/src/main/java/org/iut/nantes/factories/cours/Prof.java
package org.iut.nantes.factories.cours;

abstract class Prof
{
  protected String nom;
  protected String domaine;
  
  public String getNom() { return nom; }
  public String getDomaine() { return domaine; }

  public Prof( String n )
  {
    nom = n;
  }
}

class ProfInfo extends Prof
{
  public ProfInfo( String n )
  {
    super( n );
    domaine = "info";
  }
}

class ProfMaths extends Prof
{
  public ProfMaths( String n )
  {
    super( n );
    domaine = "maths";
  }
}

class ProfChimie extends Prof
{
  public ProfChimie( String n )
  {
    super( n );
    domaine = "chimie";
  }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/factories/cours/Prof.kt
package org.iut.nantes.factories.cours

abstract class Prof(var nom: String) {
    var domaine: String? = null
        protected set
}

class ProfInfo(n: String) : Prof(n) {
    init {
        domaine = "info"
    }
}

class ProfMaths(n: String) : Prof(n) {
    init {
        domaine = "maths"
    }
}

class ProfChimie(n: String) : Prof(n) {
    init {
        domaine = "chimie"
    }
}